75521 rustdoc book improvements (#75778)
rust-lang/rust: src/doc/rustdoc/src/how-to-write-documentation.md

requirements are outlined, the edge cases shown, and practical examples 
provided.  The front page goes on to show how to use regular expressions
then concludes with crate features.

Don't worry about comparing your crate, which is just beginning, to other more
developed crates.  To get the documentation to something more polished, start
incrementally and put in an introduction, example, and features.  Rome was not 
built in a day!

The first lines within the `lib.rs` will compose the front-page, and they
use a different convention than the rest of the rustdocs.  Lines should
start with `//!` which indicate module-level or crate-level documentation.
Here's a quick example of the difference:

```rust,ignore
//! Fast and easy queue abstraction.
//!
//! Provides an abstraction over a queue.  When the abstraction is used
//! there are these advantages:
//! - Fast
//! - [`Easy`]
//!
//! [`Easy`]: http://thatwaseasy.example.com

/// This module makes it easy.
pub mod easy {

    /// Use the abstract function to do this specific thing.
    pub fn abstract() {}

}
```

Ideally, this first line of documentation is a sentence without highly 

> technical details, but descriptive of where this crate fits

Reviewer: ```suggestion
technical details, but with a good description of where this crate fits
```

start with `//!` which indicate module-level or crate-level documentation.
Here's a quick example of the difference:

> ```rust,ignore

Reviewer: Opened https://github.com/rust-lang/rust/issues/76709; no need to change it here.

in searches and module overviews.  For example, the function `std::env::args()`
above will be shown on the [`std::env`] module documentation.  It is good
practice to keep the summary to one line: concise writing is a goal of good
documentation.

Because the type system does a good job of defining what types a function
passes and returns, there is no benefit of explicitly writing it
into the documentation, especially since `rustdoc` adds hyper links to all types in the function signature.

In the example above, a 'Panics' section explains when the code might abruptly exit,
which can help the reader prevent reaching a panic.  A panic section is recommended
every time edge cases in your code can be reached if known.

As you can see, it follows the structure detailed above: it starts with a short
sentence explaining what the functions does, then it provides more information
and finally provides a code example.

## Markdown

`rustdoc` uses the [commonmark markdown specification]. You might be
interested into taking a look at their website to see what's possible to do.
 - [commonmark quick reference]
 - [current spec]

> [backtrace]: https://docs.rs/backtrace/0.3.50/backtrace/

Reviewer: ```
rustdoc/how-to-write-documentation.html:187: broken intra-doc link - [<code>backtrace</code>]
rustdoc/print.html:306: broken intra-doc link - [<code>backtrace</code>]
thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:109:9
```
```suggestion
[`backtrace`]: https://docs.rs/backtrace/0.3.50/backtrace/
```

[`args_os`]: ./fn.args_os.html
``````

Everything before the first empty line will be reused to describe the component 
in searches and module overviews.  For example, the function `std::env::args()`
above will be shown on the [`std::env`] module documentation.  It is good
practice to keep the summary to one line: concise writing is a goal of good
documentation.

> Because the type system does a good job of defining what is passed to a function
> and what is returned from one, there is not a benefit of explicitly writing it

Reviewer: ```suggestion
Because the type system does a good job of defining what types a function
passes and returns, there is no benefit of explicitly writing it
```

> into the documentation.  Rustdoc makes sure the links to the types included
> in the function signature are linked.

Reviewer: ```suggestion
into the documentation, especially since `rustdoc` adds hyper links to all types in the function signature.
```

`rustdoc` is using the [commonmark markdown specification]. You might be
interested into taking a look at their website to see what's possible to do.
 - [commonmark quick reference]
 - [current spec]

## Lints

To be sure that you didn't miss any item without documentation or code examples,

> you can take a look at the rustdoc lints [here][rustdoc-lints].

Reviewer: Hmm, is there a reason you removed this?

Author: I gave that section an entire chapter.   I moved it into the what-to-include section, although I rewrote it to have details.

Reviewer: Oh nice! Much better :D
